Core/include/Acts/Surfaces/BoundaryTolerance.hpp

```diff
@@ -0,0 +1,154 @@
+    AbsoluteBound(double tolerance0_, double tolerance1_)
+        : tolerance0(tolerance0_), tolerance1(tolerance1_) {}
+  };
+
+  /// Absolute tolerance in Cartesian coordinates
+  struct AbsoluteCartesian {
+    double tolerance0{};
+    double tolerance1{};
+
+    AbsoluteCartesian() = default;
+    AbsoluteCartesian(double tolerance0_, double tolerance1_)
+        : tolerance0(tolerance0_), tolerance1(tolerance1_) {}
+  };
+
+  /// Absolute tolerance in Euclidean distance
+  struct AbsoluteEuclidean {
+    double tolerance{};
+
+    AbsoluteEuclidean() = default;
+    AbsoluteEuclidean(double tolerance_) : tolerance(tolerance_) {}
+  };
+
+  /// Chi2 tolerance in bound coordinates
+  struct Chi2Bound {
+    double maxChi2{};
+    SquareMatrix2 weight = SquareMatrix2::Identity();
+
+    Chi2Bound() = default;
+    Chi2Bound(const SquareMatrix2& weight_, double maxChi2_)
+        : maxChi2(maxChi2_), weight(weight_) {}
+  };
+
+  /// Underlying variant type
+  using Variant = std::variant<Infinite, None, AbsoluteBound, AbsoluteCartesian,
+                               AbsoluteEuclidean, Chi2Bound>;
+
+  /// Construct with infinite tolerance.
+  BoundaryTolerance(const Infinite& infinite);
+  /// Construct with no tolerance.
+  BoundaryTolerance(const None& none);
+  /// Construct with absolute tolerance in bound coordinates.
+  BoundaryTolerance(const AbsoluteBound& AbsoluteBound);
+  /// Construct with absolute tolerance in Cartesian coordinates.
+  BoundaryTolerance(const AbsoluteCartesian& absoluteCartesian);
+  /// Construct with absolute tolerance in Euclidean distance.
+  BoundaryTolerance(const AbsoluteEuclidean& absoluteEuclidean);
+  /// Construct with chi2 tolerance in bound coordinates.
+  BoundaryTolerance(const Chi2Bound& Chi2Bound);
+
+  /// Construct from variant
+  BoundaryTolerance(Variant variant);
+
+  /// Check if the tolerance is infinite.
+  bool isInfinite() const;
+  /// Check if the is no tolerance.
+  bool isNone() const;
+  /// Check if the tolerance is absolute with bound coordinates.
+  bool hasAbsoluteBound(bool isCartesian = false) const;
+  /// Check if the tolerance is absolute with Cartesian coordinates.
+  bool hasAbsoluteCartesian() const;
+  /// Check if the tolerance is absolute with Euclidean distance.
+  bool hasAbsoluteEuclidean() const;
+  /// Check if the tolerance is chi2 with bound coordinates.
+  bool hasChi2Bound() const;
+
+  /// Check if any tolerance is set.
+  bool hasTolerance() const;
+
+  /// Get the tolerance as absolute bound.
+  AbsoluteBound asAbsoluteBound(bool isCartesian = false) const;
+  /// Get the tolerance as absolute Cartesian.
+  const AbsoluteCartesian& asAbsoluteCartesian() const;
+  /// Get the tolerance as absolute Euclidean.
+  const AbsoluteEuclidean& asAbsoluteEuclidean() const;
+  /// Get the tolerance as chi2 bound.
+  const Chi2Bound& asChi2Bound() const;
+
+  /// Get the tolerance as absolute bound if possible.
+  std::optional<AbsoluteBound> asAbsoluteBoundOpt(
```

Comment:
This makes a copy right? If this is a performance concern (which I'm thinking it might be), we could instead return a pointer (instead of a reference) and use that to communicate the optional aspect. That's less safe though, of course.

Reply:
True that will be used in a critical path and it results in a copy. Since `AbsoluteBound` is quite small I didn't expect any big overheads from that but this is something we can try to measure.